diff -ur gettext-0.12.1-official/gettext-runtime/intl/relocatable.c gettext-0.12.1/gettext-runtime/intl/relocatable.c
--- gettext-0.12.1-official/gettext-runtime/intl/relocatable.c	Mon Apr  7 10:42:24 2003
+++ gettext-0.12.1/gettext-runtime/intl/relocatable.c	Fri Jun 20 15:17:20 2003
@@ -83,6 +83,11 @@
 # define FILESYSTEM_PREFIX_LEN(P) 0
 #endif
 
+#if defined _WIN32 || defined __WIN32__
+# define WIN32_LEAN_AND_MEAN
+# include <windows.h>
+#endif
+
 /* Original installation prefix.  */
 static char *orig_prefix;
 static size_t orig_prefix_len;
diff -ur gettext-0.12.1-official/gettext-tools/lib/Makefile.in gettext-0.12.1/gettext-tools/lib/Makefile.in
--- gettext-0.12.1-official/gettext-tools/lib/Makefile.in	Fri Jun 20 16:32:28 2003
+++ gettext-0.12.1/gettext-tools/lib/Makefile.in	Fri Jun 20 15:29:56 2003
@@ -344,7 +344,7 @@
 am_libgettextlib_la_OBJECTS = argmatch.lo backupfile.lo addext.lo \
 	basename.lo c-ctype.lo classpath.lo copy-file.lo error.lo \
 	execute.lo findprog.lo fstrcmp.lo full-write.lo gcd.lo \
-	getopt.lo getopt1.lo hash.lo javacomp.lo javaexec.lo \
+	hash.lo javacomp.lo javaexec.lo \
 	linebreak.lo mbswidth.lo obstack.lo concatpath.lo pipe-bidi.lo \
 	pipe-in.lo pipe-out.lo progname.lo progreloc.lo safe-read.lo \
 	safe-write.lo sh-quote.lo stpncpy.lo strtoul.lo tmpdir.lo \
diff -ur gettext-0.12.1-official/gettext-tools/lib/canonicalize.c gettext-0.12.1/gettext-tools/lib/canonicalize.c
--- gettext-0.12.1-official/gettext-tools/lib/canonicalize.c	Fri May  9 10:12:06 2003
+++ gettext-0.12.1/gettext-tools/lib/canonicalize.c	Fri Jun 20 15:22:38 2003
@@ -74,6 +74,10 @@
 # define __readlink readlink
 #endif
 
+#if defined _WIN32 || defined __WIN32__
+# define lstat(a,b) stat(a,b)
+#endif
+
 /* Return the canonical absolute name of file NAME.  A canonical name
    does not contain any `.', `..' components nor any repeated path
    separators ('/') or symlinks.  All path components must exist.  If
diff -ur gettext-0.12.1-official/gettext-tools/lib/mkdtemp.c gettext-0.12.1/gettext-tools/lib/mkdtemp.c
--- gettext-0.12.1-official/gettext-tools/lib/mkdtemp.c	Tue Jan 14 13:36:08 2003
+++ gettext-0.12.1/gettext-tools/lib/mkdtemp.c	Fri Jun 20 15:21:50 2003
@@ -92,6 +92,10 @@
 # define __mkdir mkdir
 #endif
 
+#if defined _WIN32 || defined __WIN32__
+# define mkdir(a,b) mkdir(a)
+#endif
+
 /* Use the widest available unsigned type if uint64_t is not
    available.  The algorithm below extracts a number less than 62**6
    (approximately 2**35.725) from uint64_t, so ancient hosts where
diff -ur gettext-0.12.1-official/gettext-tools/lib/xreadlink.c gettext-0.12.1/gettext-tools/lib/xreadlink.c
--- gettext-0.12.1-official/gettext-tools/lib/xreadlink.c	Fri Feb 28 12:19:46 2003
+++ gettext-0.12.1/gettext-tools/lib/xreadlink.c	Fri Jun 20 15:18:28 2003
@@ -60,6 +60,11 @@
 # include "xmalloc.h"
 #endif
 
+#if defined _WIN32 || defined __WIN32__
+  typedef long ssize_t;
+# define readlink(a,b,c) strncpy(b,a,c)
+#endif
+
 /* Call readlink to get the symbolic link value of FILENAME.
    Return a pointer to that NUL-terminated string in malloc'd storage.
    If readlink fails, return NULL (caller may use errno to diagnose).
diff -ur gettext-0.12.1-official/gettext-tools/src/Makefile.in gettext-0.12.1/gettext-tools/src/Makefile.in
--- gettext-0.12.1-official/gettext-tools/src/Makefile.in	Fri Jun 20 16:33:08 2003
+++ gettext-0.12.1/gettext-tools/src/Makefile.in	Fri Jun 20 15:39:52 2003
@@ -456,7 +456,7 @@
 
 am_hostname_OBJECTS = hostname-hostname.$(OBJEXT)
 hostname_OBJECTS = $(am_hostname_OBJECTS)
-hostname_LDADD = $(LDADD)
+hostname_LDADD = $(LDADD) -lwsock32
 hostname_DEPENDENCIES = ../lib/libgettextlib.la
 am_msgattrib_OBJECTS = msgattrib-msgattrib.$(OBJEXT)
 msgattrib_OBJECTS = $(am_msgattrib_OBJECTS)
diff -ur gettext-0.12.1-official/gettext-tools/src/hostname.c gettext-0.12.1/gettext-tools/src/hostname.c
--- gettext-0.12.1-official/gettext-tools/src/hostname.c	Fri Apr 25 10:18:42 2003
+++ gettext-0.12.1/gettext-tools/src/hostname.c	Fri Jun 20 15:38:42 2003
@@ -52,9 +52,13 @@
 
 /* Support for using gethostbyname().  */
 #if HAVE_GETHOSTBYNAME
-# include <sys/types.h>
-# include <sys/socket.h> /* defines AF_INET, AF_INET6 */
-# include <netinet/in.h> /* declares ntohs(), defines struct sockaddr_in */
+# if defined _WIN32 || defined __WIN32__
+#  include <winsock.h>
+# else
+#  include <sys/types.h>
+#  include <sys/socket.h> /* defines AF_INET, AF_INET6 */
+#  include <netinet/in.h> /* declares ntohs(), defines struct sockaddr_in */
+# endif
 # if HAVE_ARPA_INET_H
 #  include <arpa/inet.h> /* declares inet_ntoa(), inet_ntop() */
 # endif
@@ -71,7 +75,9 @@
 #   define HAVE_IN6_S6_ADDR16 1
 #  endif
 # endif
-# include <netdb.h> /* defines struct hostent, declares gethostbyname() */
+# if !defined _WIN32 && !defined __WIN32__
+#  include <netdb.h> /* defines struct hostent, declares gethostbyname() */
+# endif
 #endif
 
 /* Include this after <sys/socket.h>, to avoid a syntax error on BeOS.  */
@@ -238,10 +244,20 @@
 {
 #if HAVE_GETHOSTNAME
   char hostname[MAXHOSTNAMELEN+1];
+  
+# if defined _WIN32 || defined __WIN32__
+  WSADATA wsaData;
+  WSAStartup(MAKEWORD(1,1),&wsaData);
+# endif
 
   if (gethostname (hostname, MAXHOSTNAMELEN) < 0)
     error (EXIT_FAILURE, errno, _("could not get host name"));
   hostname[MAXHOSTNAMELEN] = '\0';
+
+# if defined _WIN32 || defined __WIN32__
+  WSACleanup();
+# endif
+
   return xstrdup (hostname);
 #else
   struct utsname utsname;
diff -ur gettext-0.12.1-official/gettext-tools/src/write-java.c gettext-0.12.1/gettext-tools/src/write-java.c
--- gettext-0.12.1-official/gettext-tools/src/write-java.c	Tue Apr 29 10:04:04 2003
+++ gettext-0.12.1/gettext-tools/src/write-java.c	Fri Jun 20 15:32:20 2003
@@ -80,6 +80,9 @@
 
 #define _(str) gettext (str)
 
+#if defined _WIN32 || defined __WIN32__
+# define mkdir(a,b) mkdir(a)
+#endif
 
 /* Check that the resource name is a valid Java class name.  To simplify
    things, we allow only ASCII characters in the class name.
