diff -ur libiconv-1.9.1-official/lib/relocatable.c libiconv-1.9.1/lib/relocatable.c
--- libiconv-1.9.1-official/lib/relocatable.c	Sat Apr  5 10:03:26 2003
+++ libiconv-1.9.1/lib/relocatable.c	Fri Jun 20 14:51:50 2003
@@ -83,6 +83,11 @@
 # define FILESYSTEM_PREFIX_LEN(P) 0
 #endif
 
+#if defined _WIN32 || defined __WIN32__
+# define WIN32_LEAN_AND_MEAN
+# include <windows.h>
+#endif
+
 /* Original installation prefix.  */
 static char *orig_prefix;
 static size_t orig_prefix_len;
diff -ur libiconv-1.9.1-official/libcharset/lib/relocatable.c libiconv-1.9.1/libcharset/lib/relocatable.c
--- libiconv-1.9.1-official/libcharset/lib/relocatable.c	Sat Apr  5 10:03:26 2003
+++ libiconv-1.9.1/libcharset/lib/relocatable.c	Fri Jun 20 14:51:34 2003
@@ -83,6 +83,11 @@
 # define FILESYSTEM_PREFIX_LEN(P) 0
 #endif
 
+#if defined _WIN32 || defined __WIN32__
+# define WIN32_LEAN_AND_MEAN
+# include <windows.h>
+#endif
+
 /* Original installation prefix.  */
 static char *orig_prefix;
 static size_t orig_prefix_len;
diff -ur libiconv-1.9.1-official/srclib/xreadlink.c libiconv-1.9.1/srclib/xreadlink.c
--- libiconv-1.9.1-official/srclib/xreadlink.c	Fri Feb 28 12:19:46 2003
+++ libiconv-1.9.1/srclib/xreadlink.c	Fri Jun 20 14:52:28 2003
@@ -60,6 +60,11 @@
 # include "xmalloc.h"
 #endif
 
+#if defined _WIN32 || defined __WIN32__
+  typedef long ssize_t;
+# define readlink(a,b,c) strncpy(b,a,c)
+#endif
+
 /* Call readlink to get the symbolic link value of FILENAME.
    Return a pointer to that NUL-terminated string in malloc'd storage.
    If readlink fails, return NULL (caller may use errno to diagnose).
